### PR TITLE
fix: Fix crash when nested CFN stack has dict TemplateURL (unresolved intrinsics)

### PR DESCRIPTION
**Pull request by aahung**
#### Which issue(s) does this change fix?

#2832

#### Why is this change necessary?

This PR does not fix the use case in #2832, but rather make SAM CLI handle well when nested CFN stack's `TemplateURL` is a dictionary (pointing to a stack on SAR). This makes it consistent with how we handle nested serverless applications:
https://github.com/aws/aws-sam-cli/blob/9e03633dc6f8c5bdaef3b88be67314f00c28d10d/samcli/lib/providers/sam_stack_provider.py#L128-L145

#### Checklist

- [ ] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [ ] Write unit tests
- [ ] Write/update functional tests
- [ ] Write/update integration tests
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/lib/providers/sam_stack_provider.py [hunk 1]
```diff
@@ -163,8 +163,14 @@ def _convert_cfn_stack_resource(
         resource_properties: Dict,
         global_parameter_overrides: Optional[Dict] = None,
     ) -> Optional[Stack]:
-        template_url = resource_properties.get("TemplateURL", "")
+        template_url = resource_properties.get("TemplateURL")
 
+        if isinstance(template_url, dict):
+            # This happens when TemplateURL has unresolvable intrinsic functions
+            # and it usually happens in CDK generated template files (#2832).
```
Comment (hawflau) [inline samcli/lib/providers/sam_stack_provider.py:170]: wondering if we can show a message to tell customers to try `sam-beta-cdk` ([link](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-cdk.html)

Comment (aahung) [reply samcli/lib/providers/sam_stack_provider.py:170]: Would that solve this problem?

File: samcli/lib/providers/sam_stack_provider.py [hunk 2]
```diff
@@ -163,8 +163,14 @@ def _convert_cfn_stack_resource(
         resource_properties: Dict,
         global_parameter_overrides: Optional[Dict] = None,
     ) -> Optional[Stack]:
-        template_url = resource_properties.get("TemplateURL", "")
+        template_url = resource_properties.get("TemplateURL")
 
+        if isinstance(template_url, dict):
+            # This happens when TemplateURL has unresolvable intrinsic functions
```
Comment (hoffa) [inline samcli/lib/providers/sam_stack_provider.py:169]: > This happens when TemplateURL has unresolvable intrinsic functions

Why?

Comment (aahung) [reply samcli/lib/providers/sam_stack_provider.py:169]: unresolved intrinsic functions have type "dict"

**Review by hoffa — APPROVED** [review]

**Review by hawflau — APPROVED** [review]

**Pull request merged**